distributed/scheduler.py

```diff
@@ -1804,30 +1818,64 @@ def decide_worker(self, ts: TaskState) -> WorkerState | None:
             and len(tg.dependencies) < 5
             and sum(map(len, tg.dependencies)) < 5
         ):
-            ws = tg.last_worker
-
-            if not (ws and tg.last_worker_tasks_left and ws.address in self.workers):
-                # Last-used worker is full or unknown; pick a new worker for the next few tasks
-                ws = min(
-                    (self.idle or self.workers).values(),
-                    key=partial(self.worker_objective, ts),
-                )
-                assert ws
-                tg.last_worker_tasks_left = math.floor(
-                    (len(tg) / self.total_nthreads) * ws.nthreads
-                )
-
-            # Record `last_worker`, or clear it on the final task
-            tg.last_worker = (
-                ws if tg.states["released"] + tg.states["waiting"] > 1 else None
+            if math.isinf(self.WORKER_SATURATION):  # no scheduler-side queuing
+                pool = self.idle.values() if self.idle else self.running
+                if not pool:
+                    recommendations[ts.key] = "no-worker"
+                    return None
+
+                lws = self.last_root_worker
+                if not (
+                    lws
+                    and self.last_root_worker_tasks_left
+                    and self.workers.get(lws.address) is lws
+                ):
+                    # Last-used worker is full or unknown; pick a new worker for the next few tasks
+                    ws = self.last_root_worker = min(
+                        pool, key=lambda ws: len(ws.processing) / ws.nthreads
+                    )
```

Comment:
Good question. Before this, we used `worker_objective` which considers both occupancy and data transfer. Since these are root tasks, data transfer is irrelevant, so this seemed more direct. But there's probably not much reason to look at `len(ws.processing)` instead of `ws.occupancy`—you're right that they should be about the same.

I guess one reason would be that we decide when the worker is full based on the number of tasks, not the occupancy, so maybe it makes sense to judge the "most empty" one by the same metric? I don't think that'll matter too much in this overproduction case, but maybe it makes more sense for the queuing case below.